File: shared/src/asistente/manifiesto.ts
import { z } from 'zod';

/**
 * Manifiesto del asistente: el contrato entre el backend generado y la app
 * móvil.
 *
 * El problema que resuelve. El backend que sale del generador es distinto para
 * cada diagrama: hoy tiene `Cita` y `Barbero`, mañana `Pedido` y `Producto`. La
 * app móvil que lo consume por voz no puede conocer esos nombres de antemano, y
 * generar un proyecto Flutter por diagrama significaría recompilar e instalar un
 * APK cada vez que alguien mueve una caja en el lienzo. Eso no es una
 * herramienta colaborativa: es un ciclo de despliegue de veinte minutos metido
 * en medio de una clase de diseño.
 *
 * La salida es invertir la dependencia. La app se compila **una vez** y no sabe
 * nada del dominio; al conectarse pide este manifiesto y de él deriva todo: qué
 * entidades existen, qué campos tiene cada una, de qué tipo son, cuáles son
 * obligatorios, qué acciones se pueden ejecutar y cuáles hay que confirmar en
 * voz alta antes de hacerlas. Los formularios y la gramática de órdenes se
 * construyen en tiempo de ejecución a partir de estos datos.
 *
 * Por qué vive en `shared` y no en el generador. Lo escriben dos partes —el
 * emisor del generador y el lector de la app— y lo validan las pruebas de
 * ambas. Un contrato que solo conoce quien lo emite no es un contrato.
 *
 * Lo que deliberadamente NO lleva:
 *
 * - **Fecha de generación.** Haría que dos ejecuciones del generador sobre el
 *   mismo diagrama produjeran ficheros distintos, y con ello se pierde la
 *   propiedad más útil que tiene un generador: que puedas volver a lanzarlo y
 *   comparar. La versión del esquema y el nombre del proyecto bastan.
 * - **Credenciales ni URLs absolutas.** La app ya sabe a qué servidor se ha
 *   conectado. Meter aquí un host lo congelaría al primero que se usó, que es
 *   justo lo que rompe al pasar de localhost a la nube.
 */

/**
 * Versión del esquema. Se compara con igualdad, no con «mayor o igual»: es
 * preferible que la app diga «este servidor habla una versión que no entiendo»
 * a que dibuje media pantalla con los campos que sí reconoce y silencie los
 * demás. Un formulario al que le falta un campo obligatorio en silencio produce
 * un 400 que el usuario no sabe explicar.
 */
export const VERSION_MANIFIESTO = 1;

/**
 * Tipos de campo tal y como los entiende la app, no como los entiende Java.
 *
 * Es una traducción con pérdida y a propósito. A la app le da igual que algo
 * sea `Integer` o `Long`: lo que necesita saber es qué teclado abrir, cómo
 * validar antes de enviar y cómo leer el valor en voz alta. `BigDecimal` y
 * `Double` colapsan en `decimal` porque el diálogo hablado es el mismo.
 */
export const TipoCampoSchema = z.enum([
  'texto',
  'entero',
  'decimal',
  'booleano',
  'fecha',
  'fechaHora',
  'hora',
  'uuid',
  /** Valor de un conjunto cerrado: el manifiesto trae la lista en `valores`. */
  'enumerado',
  /** Clave ajena a otra entidad del manifiesto, nombrada en `entidad`. */
  'referencia',
]);
export type TipoCampo = z.infer<typeof TipoCampoSchema>;

/**
 * Las cinco operaciones que el controlador generado expone hoy. Se enumeran de
 * forma explícita por entidad en vez de darlas por supuestas: en cuanto una
 * entidad deje de exponer el borrado —una entidad de auditoría, por ejemplo— la
 * app tiene que dejar de ofrecerlo sin que haya que tocarla.
 */
export const AccionSchema = z.enum(['listar', 'ver', 'crear', 'actualizar', 'borrar']);
export type Accion = z.infer<typeof AccionSchema>;

export const CampoManifiestoSchema = z.object({
  /** Nombre de la propiedad JSON tal cual la espera el DTO de entrada. */
  nombre: z.string().min(1).max(64),
  /**
   * El mismo nombre en palabras separadas y en minúsculas: `precioUnitario` se
   * convierte en `precio unitario`. Es lo que se muestra en la etiqueta del
   * formulario y lo que se pronuncia; también es contra lo que se compara la
   * frase dictada, porque nadie dicta en camelCase.
   */
  etiqueta: z.string().min(1).max(128),
  tipo: TipoCampoSchema,
  obligatorio: z.boolean(),
  /** El identificador: se muestra pero no se envía ni se dicta. */
  soloLectura: z.boolean(),
  /** Longitud máxima declarada por el `@Size` del DTO, si la hay. */
  maxLongitud: z.number().int().positive().optional(),
  /** Solo en `enumerado`: el conjunto cerrado de valores admitidos. */
  valores: z.array(z.string().min(1)).optional(),
  /** Solo en `referencia`: el `nombre` de la entidad apuntada. */
  entidad: z.string().min(1).max(64).optional(),
});
export type CampoManifiesto = z.infer<typeof CampoManifiestoSchema>;

export const EntidadManifiestoSchema = z.object({
  /** Nombre de la clase del diagrama, en PascalCase. */
  nombre: z.string().min(1).max(64),
  /**
   * Segmento de ruta que se concatena a `baseUrl`, ya en plural y kebab-case,
   * con la barra inicial: `/linea-pedidos`.
   *
   * El patrón es restrictivo a propósito. Este valor acaba formando una URL, y
   * viene de un nombre que ha escrito una persona dibujando un diagrama
   * (RNF-SEG-06). Una entidad llamada `../admin` no debe poder convertirse en
   * una petición a otro sitio.
   */
  ruta: z
    .string()
    .regex(/^\/[a-z0-9]+(-[a-z0-9]+)*$/, 'la ruta debe ser kebab-case con barra inicial'),
  /**
   * Cómo se nombra la entidad al hablar. Se emiten las formas que pueden
   * derivarse del diagrama —singular y plural, sin tildes y en minúsculas— y
   * nada más: inventar sinónimos que el usuario no ha escrito llevaría a que la
   * app reconociera «turno» en un proyecto donde ese nombre significa otra
   * cosa. Los sinónimos propios del dominio son una extensión posterior, y su
   * sitio natural es el diagrama, no el generador.
   */
  etiquetaHablada: z.array(z.string().min(1)).min(1),
  identificador: z.object({
    nombre: z.string().min(1).max(64),
    tipo: TipoCampoSchema,
  }),
  /**
   * Campo que representa al registro cuando hay que nombrarlo: el primer campo
   * de texto obligatorio. Permite decir «la cita de Juan» en lugar de «la cita
   * número 47», y ofrecer una lista legible al elegir una referencia.
   * Ausente si la entidad no tiene ningún campo de texto.
   */
  campoEtiqueta: z.string().min(1).max(64).optional(),
  campos: z.array(CampoManifiestoSchema),
  acciones: z.array(AccionSchema),
  /**
   * Acciones que exigen confirmación hablada antes de ejecutarse (RF-VOZ-04).
   * Una orden mal reconocida que crea un registro de más es un incordio; una
   * que borra uno es una pérdida de datos.
   */
  criticas: z.array(AccionSchema),
  /**
   * Entidades que desaparecen si se borra esta, por ser partes de una
   * composición con cascada.
   *
   * Está aquí porque el aviso hablado tiene que ser honesto: «borrar este
   * pedido eliminará también sus líneas» es una advertencia distinta de «¿borro
   * este pedido?», y la diferencia solo puede saberla quien ha visto el rombo
   * relleno en el diagrama. La app no puede deducirlo de la API REST.
   */
  borradoEnCascada: z.array(z.string().min(1)),
});
export type EntidadManifiesto = z.infer<typeof EntidadManifiestoSchema>;

export const EnumeradoManifiestoSchema = z.object({
  nombre: z.string().min(1).max(64),
  valores: z.array(z.string().min(1)),
});
export type EnumeradoManifiesto = z.infer<typeof EnumeradoManifiestoSchema>;

export const ManifiestoSchema = z.object({
  version: z.literal(VERSION_MANIFIESTO),
  /** Nombre legible del proyecto; se usa al presentarse por voz. */
  proyecto: z.string().min(1).max(128),
  /** Prefijo común de todos los controladores generados. */
  baseUrl: z.string().regex(/^\/[a-z0-9/-]*$/),
  /**
   * Idioma de las etiquetas y de la gramática. Hoy solo hay castellano; el
   * campo existe para que la app pueda rechazar un manifiesto que no sabe
   * pronunciar en vez de leerlo con la voz equivocada.
   */
  idioma: z.literal('es'),
  /**
   * Nombre de la cabecera de idempotencia que acepta este backend. Va en el
   * manifiesto y no está fijado en la app porque es el servidor quien decide si
   * la implementa: si el campo falta, la app sabe que reenviar es peligroso y
   * debe pedir confirmación en vez de reintentar sola.
   */
  cabeceraIdempotencia: z.string().min(1).max(64).optional(),
  entidades: z.array(EntidadManifiestoSchema),
  enumerados: z.array(EnumeradoManifiestoSchema),
});
export type Manifiesto = z.infer<typeof ManifiestoSchema>;

/** Cabecera que emiten el generador y la app. Un solo sitio donde cambiarla. */
export const CABECERA_IDEMPOTENCIA = 'Idempotency-Key';

/**
 * Busca una entidad por cualquiera de sus formas habladas.
 *
 * Vive aquí, junto al esquema, porque lo necesitan las dos orillas: el
 * intérprete de órdenes del móvil y las pruebas del generador que comprueban
 * que las frases del corpus encuentran a quién se refieren. Duplicarlo sería
 * garantizar que las dos copias se separen.
 */
export function buscarEntidadHablada(
  manifiesto: Manifiesto,
  termino: string,
): EntidadManifiesto | undefined {
  const buscado = plegar(termino);
  return manifiesto.entidades.find((entidad) =>
    entidad.etiquetaHablada.some((etiqueta) => plegar(etiqueta) === buscado),
  );
}

/** Igual que `buscarEntidadHablada`, para un campo dentro de una entidad. */
export function buscarCampoHablado(
  entidad: EntidadManifiesto,
  termino: string,
): CampoManifiesto | undefined {
  const buscado = plegar(termino);
  return entidad.campos.find(
    (campo) => plegar(campo.etiqueta) === buscado || plegar(campo.nombre) === buscado,
  );
}

/**
 * Normaliza para comparar lo dictado con lo escrito: minúsculas, sin tildes y
 * con los espacios colapsados. Es la misma regla que aplica la gramática de
 * `ai/grammar.ts`; se repite en cuatro líneas en vez de exportarla desde allí
 * para no acoplar el contrato del manifiesto al intérprete del lienzo, que
 * evoluciona por otros motivos.
 */
const DIACRITICOS = new RegExp('[\\u0300-\\u036f]', 'g');

function plegar(texto: string): string {
  return texto
    .toLowerCase()
    .normalize('NFD')
    .replace(DIACRITICOS, '')
    .trim()
    .replace(/\s+/g, ' ');
}

File: mobile/lib/asistente/manifiesto.dart
/// El manifiesto tal y como lo entiende la app.
///
/// Es la otra mitad del contrato que emite el generador en
/// `generator/src/asistente/manifiesto.ts` y que define
/// `shared/src/asistente/manifiesto.ts`. Que existan dos implementaciones —una
/// en TypeScript, otra en Dart— no es duplicación evitable: son dos runtimes
/// distintos, y el precio de tener una app sola para todos los proyectos
/// generados es escribir el lector una vez en cada lado.
///
/// Lo que **no** se hace aquí es confiar. La URL del backend la teclea el
/// usuario en la pantalla de conexión, así que el documento que llega es
/// entrada externa como cualquier otra: sus rutas acaban concatenadas en una
/// URL y sus nombres acaban en la interfaz. Se validan con lista blanca —la
/// misma que aplica el emisor— y se rechaza el documento entero si algo no
/// encaja. Nunca se «limpia» quitando caracteres: una ruta saneada a medias
/// sigue siendo una ruta que nadie escribió.
library;

/// Versión del contrato que esta app sabe leer.
///
/// Se compara con igualdad y no con «mayor o igual» a propósito. Un manifiesto
/// de la versión 2 puede haber cambiado el significado de un campo existente,
/// no solo añadido campos nuevos; seguir adelante sería adivinar. El mensaje de
/// error dice qué versión habla cada lado, que es lo único accionable.
const int versionManifiesto = 1;

/// Cabecera de idempotencia por defecto, si el manifiesto no anuncia otra.
const String cabeceraIdempotenciaPorDefecto = 'Idempotency-Key';

/// El documento no es válido y no se puede seguir.
class ManifiestoInvalido implements Exception {
  const ManifiestoInvalido(this.mensaje);

  final String mensaje;

  @override
  String toString() => 'Manifiesto inválido: $mensaje';
}

/// Tipos de campo del contrato. El orden es el de `TipoCampoSchema`.
enum TipoCampo {
  texto,
  entero,
  decimal,
  booleano,
  fecha,
  fechaHora,
  hora,
  uuid,
  enumerado,
  referencia;

  static TipoCampo desde(String valor, String donde) {
    for (final tipo in TipoCampo.values) {
      if (tipo.name == valor) return tipo;
    }
    // Se rechaza en vez de caer a `texto`. Como la versión está clavada en 1, un
    // tipo desconocido solo puede venir de un backend incoherente, y mandarle un
    // texto donde espera un número convierte un fallo ruidoso en uno callado.
    throw ManifiestoInvalido('$donde: tipo «$valor» desconocido');
  }
}

/// Acciones que una entidad admite.
enum Accion {
  listar,
  ver,
  crear,
  actualizar,
  borrar;

  static Accion desde(String valor, String donde) {
    for (final accion in Accion.values) {
      if (accion.name == valor) return accion;
    }
    throw ManifiestoInvalido('$donde: acción «$valor» desconocida');
  }
}

/// Un campo de una entidad, tal y como se dicta y se rellena.
class CampoManifiesto {
  const CampoManifiesto({
    required this.nombre,
    required this.etiqueta,
    required this.tipo,
    required this.obligatorio,
    required this.soloLectura,
    this.maxLongitud,
    this.valores = const <String>[],
    this.entidad,
  });

  /// Nombre exacto en el DTO de entrada. Es la clave del JSON que se envía.
  final String nombre;

  /// Cómo se lee y se dice: `precioUnitario` llega aquí como «precio unitario».
  final String etiqueta;

  final TipoCampo tipo;
  final bool obligatorio;
  final bool soloLectura;

  /// Tope de `@Size(max = n)`, cuando la entidad lo declara.
  final int? maxLongitud;

  /// Lista cerrada de valores, solo para [TipoCampo.enumerado].
  ///
  /// Esto es lo que ninguna API REST cuenta de sí misma y que el diagrama sí
  /// sabe: sin ella el asistente no puede corregir «pedido enviado» a
  /// `ENVIADO`, solo puede mandarlo y esperar un 400.
  final List<String> valores;

  /// Entidad apuntada, solo para [TipoCampo.referencia].
  final String? entidad;

  factory CampoManifiesto.desdeJson(Map<String, dynamic> json, String donde) {
    final nombre = _cadena(json, 'nombre', donde);
    _exigirIdentificador(nombre, '$donde: nombre de campo');
    final tipo = TipoCampo.desde(_cadena(json, 'tipo', donde), '$donde/$nombre');

    final valores = _listaDeCadenas(json['valores'], '$donde/$nombre.valores');
    if (tipo == TipoCampo.enumerado && valores.isEmpty) {
      throw ManifiestoInvalido(
        '$donde/$nombre: un enumerado sin valores no se puede rellenar',
      );
    }

    final entidad = json['entidad'];
    if (tipo == TipoCampo.referencia && entidad is! String) {
      throw ManifiestoInvalido(
        '$donde/$nombre: una referencia debe decir a qué entidad apunta',
      );
    }
    if (entidad is String) {
      _exigirIdentificador(entidad, '$donde/$nombre: entidad apuntada');
    }

    final maximo = json['maxLongitud'];
    if (maximo != null && (maximo is! int || maximo <= 0)) {
      throw ManifiestoInvalido('$donde/$nombre: maxLongitud debe ser positiva');
    }

    return CampoManifiesto(
      nombre: nombre,
      etiqueta: _cadena(json, 'etiqueta', '$donde/$nombre'),
      tipo: tipo,
      obligatorio: _booleano(json, 'obligatorio', '$donde/$nombre'),
      soloLectura: _booleano(json, 'soloLectura', '$donde/$nombre'),
      maxLongitud: maximo as int?,
      valores: valores,
      entidad: entidad as String?,
    );
  }
}

/// El identificador de una entidad: cómo se llama y de qué tipo es.
class IdentificadorManifiesto {
  const IdentificadorManifiesto({required this.nombre, required this.tipo});

  final String nombre;
  final TipoCampo tipo;

  factory IdentificadorManifiesto.desdeJson(
    Map<String, dynamic> json,
    String donde,
  ) {
    final nombre = _cadena(json, 'nombre', donde);
    _exigirIdentificador(nombre, '$donde: identificador');
    return IdentificadorManifiesto(
      nombre: nombre,
      tipo: TipoCampo.desde(_cadena(json, 'tipo', donde), '$donde/$nombre'),
    );
  }
}

/// Una entidad: una pantalla de lista, un formulario y un puñado de órdenes.
class EntidadManifiesto {
  const EntidadManifiesto({
    required this.nombre,
    required this.ruta,
    required this.etiquetaHablada,
    required this.identificador,
    required this.campos,
    required this.acciones,
    required this.criticas,
    required this.borradoEnCascada,
    this.campoEtiqueta,
  });

  /// Nombre de la clase del diagrama. Es el que se enseña como título.
  final String nombre;

  /// Ruta REST relativa a `baseUrl`, en kebab-case y con barra inicial.
  final String ruta;

  /// Cómo se nombra al hablar: singular primero, plural después.
  ///
  /// El plural sale de [ruta] y no de una segunda pluralización, justamente
  /// para que no puedan divergir. Si se pluralizara aquí otra vez, la app
  /// acabaría pidiendo `/citas` mientras dice «citaes» y no habría manera de
  /// saber cuál de las dos está mal.
  final List<String> etiquetaHablada;

  final IdentificadorManifiesto identificador;

  /// Campo que representa al registro en una lista. Puede no haberlo.
  final String? campoEtiqueta;

  final List<CampoManifiesto> campos;
  final List<Accion> acciones;

  /// Acciones que hay que confirmar antes de ejecutar.
  final List<Accion> criticas;

  /// Qué otras entidades desaparecen al borrar una de esta.
  ///
  /// Sale del rombo relleno de una composición y **no se puede deducir desde
  /// REST**: es la diferencia entre avisar «esto también borrará sus líneas» y
  /// soltar un «¿seguro?» que no informa de nada.
  final List<String> borradoEnCascada;

  bool admite(Accion accion) => acciones.contains(accion);
  bool esCritica(Accion accion) => criticas.contains(accion);

  /// Singular para hablar y para titular una ficha.
  String get singular => etiquetaHablada.first;

  /// Plural; si el manifiesto solo trajo una forma, se reutiliza.
  String get plural => etiquetaHablada.length > 1
      ? etiquetaHablada[1]
      : etiquetaHablada.first;

  factory EntidadManifiesto.desdeJson(Map<String, dynamic> json) {
    final nombre = _cadena(json, 'nombre', 'entidad');
    _exigirIdentificador(nombre, 'entidad: nombre');

    final ruta = _cadena(json, 'ruta', nombre);
    _exigirRutaSegura(ruta, nombre);

    final habladas = _listaDeCadenas(json['etiquetaHablada'], '$nombre.etiquetaHablada');
    if (habladas.isEmpty) {
      throw ManifiestoInvalido('$nombre: sin etiqueta hablada no se puede dictar');
    }

    final campos = _lista(json['campos'], '$nombre.campos')
        .map((c) => CampoManifiesto.desdeJson(c, nombre))
        .toList(growable: false);

    final etiqueta = json['campoEtiqueta'];
    if (etiqueta != null) {
      if (etiqueta is! String) {
        throw ManifiestoInvalido('$nombre: campoEtiqueta debe ser texto');
      }
      if (!campos.any((c) => c.nombre == etiqueta)) {
        throw ManifiestoInvalido(
          '$nombre: campoEtiqueta «$etiqueta» no está entre los campos',
        );
      }
    }

    final acciones = _listaDeCadenas(json['acciones'], '$nombre.acciones')
        .map((a) => Accion.desde(a, nombre))
        .toList(growable: false);
    final criticas = _listaDeCadenas(json['criticas'], '$nombre.criticas')
        .map((a) => Accion.desde(a, nombre))
        .toList(growable: false);

    for (final critica in criticas) {
      if (!acciones.contains(critica)) {
        throw ManifiestoInvalido(
          '$nombre: «${critica.name}» está marcada como crítica pero no se ofrece',
        );
      }
    }

    final cascada = _listaDeCadenas(json['borradoEnCascada'], '$nombre.borradoEnCascada');
    for (final apuntada in cascada) {
      _exigirIdentificador(apuntada, '$nombre.borradoEnCascada');
    }

    return EntidadManifiesto(
      nombre: nombre,
      ruta: ruta,
      etiquetaHablada: habladas,
      identificador: IdentificadorManifiesto.desdeJson(
        _objeto(json['identificador'], '$nombre.identificador'),
        nombre,
      ),
      campoEtiqueta: etiqueta as String?,
      campos: campos,
      acciones: acciones,
      criticas: criticas,
      borradoEnCascada: cascada,
    );
  }

  /// Campos que el formulario debe pedir: todos menos los de solo lectura.
  List<CampoManifiesto> get camposEditables =>
      campos.where((c) => !c.soloLectura).toList(growable: false);

  CampoManifiesto? campoPorNombre(String nombre) {
    for (final campo in campos) {
      if (campo.nombre == nombre) return campo;
    }
    return null;
  }
}

/// Un enumerado del diagrama, con su lista cerrada de valores.
class EnumeradoManifiesto {
  const EnumeradoManifiesto({required this.nombre, required this.valores});

  final String nombre;
  final List<String> valores;

  factory EnumeradoManifiesto.desdeJson(Map<String, dynamic> json) {
    final nombre = _cadena(json, 'nombre', 'enumerado');
    _exigirIdentificador(nombre, 'enumerado: nombre');
    final valores = _listaDeCadenas(json['valores'], '$nombre.valores');
    if (valores.isEmpty) {
      throw ManifiestoInvalido('$nombre: un enumerado sin valores no sirve');
    }
    return EnumeradoManifiesto(nombre: nombre, valores: valores);
  }
}

/// El documento completo.
class Manifiesto {
  const Manifiesto({
    required this.version,
    required this.proyecto,
    required this.baseUrl,
    required this.idioma,
    required this.entidades,
    required this.enumerados,
    required this.cabeceraIdempotencia,
  });

  final int version;
  final String proyecto;
  final String baseUrl;
  final String idioma;
  final List<EntidadManifiesto> entidades;
  final List<EnumeradoManifiesto> enumerados;

  /// Cabecera con la que reenviar sin duplicar. Puede faltar: entonces el
  /// backend no la implementa y la bandeja de salida no debe mandarla.
  final String? cabeceraIdempotencia;

  factory Manifiesto.desdeJson(Map<String, dynamic> json) {
    final version = json['version'];
    if (version is! int) {
      throw const ManifiestoInvalido('falta el número de versión');
    }
    if (version != versionManifiesto) {
      throw ManifiestoInvalido(
        'esta app entiende la versión $versionManifiesto y el backend habla la '
        '$version. Actualiza la que se haya quedado atrás.',
      );
    }

    final baseUrl = _cadena(json, 'baseUrl', 'manifiesto');
    if (!RegExp(r'^/[a-z0-9/-]*$').hasMatch(baseUrl)) {
      throw ManifiestoInvalido('baseUrl no segura: «$baseUrl»');
    }

    final entidades = _lista(json['entidades'], 'entidades')
        .map(EntidadManifiesto.desdeJson)
        .toList(growable: false);

    final nombres = entidades.map((e) => e.nombre).toSet();
    if (nombres.length != entidades.length) {
      throw const ManifiestoInvalido('hay dos entidades con el mismo nombre');
    }
    // Una referencia a una entidad que no viene en el documento dejaría el
    // selector del formulario sin nada que ofrecer, y el fallo aparecería al
    // tocar el campo y no al conectar. Mejor ahora.
    for (final entidad in entidades) {
      for (final campo in entidad.campos) {
        if (campo.entidad != null && !nombres.contains(campo.entidad)) {
          throw ManifiestoInvalido(
            '${entidad.nombre}/${campo.nombre}: apunta a «${campo.entidad}», '
            'que no está en el manifiesto',
          );
        }
      }
    }

    final cabecera = json['cabeceraIdempotencia'];
    if (cabecera != null && cabecera is! String) {
      throw const ManifiestoInvalido('cabeceraIdempotencia debe ser texto');
    }

    return Manifiesto(
      version: version,
      proyecto: _cadena(json, 'proyecto', 'manifiesto'),
      baseUrl: baseUrl,
      idioma: _cadena(json, 'idioma', 'manifiesto'),
      entidades: entidades,
      enumerados: _lista(json['enumerados'], 'enumerados')
          .map(EnumeradoManifiesto.desdeJson)
          .toList(growable: false),
      cabeceraIdempotencia: cabecera as String?,
    );
  }

  EntidadManifiesto? entidadPorNombre(String nombre) {
    for (final entidad in entidades) {
      if (entidad.nombre == nombre) return entidad;
    }
    return null;
  }

  /// Busca la entidad por cualquiera de sus formas habladas.
  ///
  /// Compara sin acentos y sin mayúsculas porque quien dicta no las pronuncia:
  /// «categoría» y «categoria» tienen que llegar al mismo sitio.
  EntidadManifiesto? buscarHablada(String termino) {
    final buscado = plegar(termino);
    if (buscado.isEmpty) return null;
    for (final entidad in entidades) {
      for (final etiqueta in entidad.etiquetaHablada) {
        if (plegar(etiqueta) == buscado) return entidad;
      }
      if (plegar(entidad.nombre) == buscado) return entidad;
    }
    return null;
  }
}

/// Normaliza para comparar lo dictado: minúsculas, sin acentos, sin dobles
/// espacios. Se escribe el rango de diacríticos con `\u` y no con los
/// caracteres literales para que ninguna recodificación del fichero lo rompa.
String plegar(String texto) {
  final sinAcentos = texto.toLowerCase().split('').map((c) {
    const origen = 'áàäâãéèëêíìïîóòöôõúùüûñçÁÀÄÂÃÉÈËÊÍÌÏÎÓÒÖÔÕÚÙÜÛÑÇ';
    const destino = 'aaaaaeeeeiiiiooooouuuuncaaaaaeeeeiiiiooooouuuunc';
    final indice = origen.indexOf(c);
    return indice >= 0 ? destino[indice] : c;
  }).join();
  return sinAcentos.replaceAll(RegExp(r'\s+'), ' ').trim();
}

// ---------------------------------------------------------------------------
// Lectura defensiva. Todo lo de aquí abajo existe para que un documento raro
// produzca un mensaje que diga dónde está el problema, y no un
// `type 'Null' is not a subtype of type 'String'` a mitad de pantalla.

/// Mismo patrón que `RUTA_SEGURA` en el emisor. La ruta se concatena a una URL,
/// así que un `..` o una barra de más aquí es una petición a otro sitio.
final RegExp _rutaSegura = RegExp(r'^/[a-z0-9]+(-[a-z0-9]+)*$');

/// Los nombres acaban en claves de JSON y en texto de pantalla. Lista blanca.
final RegExp _identificador = RegExp(r'^[A-Za-z_][A-Za-z0-9_]*$');

void _exigirRutaSegura(String ruta, String donde) {
  if (!_rutaSegura.hasMatch(ruta)) {
    throw ManifiestoInvalido('$donde: ruta no segura «$ruta»');
  }
}

void _exigirIdentificador(String nombre, String donde) {
  if (nombre.length > 64 || !_identificador.hasMatch(nombre)) {
    throw ManifiestoInvalido('$donde: nombre no válido «$nombre»');
  }
}

String _cadena(Map<String, dynamic> json, String clave, String donde) {
  final valor = json[clave];
  if (valor is! String || valor.isEmpty) {
    throw ManifiestoInvalido('$donde: falta «$clave»');
  }
  return valor;
}

bool _booleano(Map<String, dynamic> json, String clave, String donde) {
  final valor = json[clave];
  if (valor is! bool) {
    throw ManifiestoInvalido('$donde: «$clave» debe ser verdadero o falso');
  }
  return valor;
}

Map<String, dynamic> _objeto(Object? valor, String donde) {
  if (valor is! Map<String, dynamic>) {
    throw ManifiestoInvalido('$donde: se esperaba un objeto');
  }
  return valor;
}

List<Map<String, dynamic>> _lista(Object? valor, String donde) {
  if (valor is! List) {
    throw ManifiestoInvalido('$donde: se esperaba una lista');
  }
  return valor
      .map((e) => _objeto(e, donde))
      .toList(growable: false);
}

List<String> _listaDeCadenas(Object? valor, String donde) {
  if (valor == null) return const <String>[];
  if (valor is! List) {
    throw ManifiestoInvalido('$donde: se esperaba una lista');
  }
  return valor.map((e) {
    if (e is! String || e.isEmpty) {
      throw ManifiestoInvalido('$donde: todos los elementos deben ser texto');
    }
    return e;
  }).toList(growable: false);
}
